File: IT/Programmazione funzionale/Esercizi/Python/esercizio20.py
# file esercizio20.py

from __future__ import print_function

# funzione "successore"

s = lambda n : n + 1

# restituisce il successore del successore di un numero

ff = lambda f, n : f(f(n))

# stampa "2"

print("Successore del successore di zero: {}".format(ff(s, 0)))

File: IT/Programmazione funzionale/Esercizi/C++/esercizio20.cpp
/*
file esercizio20.cpp
compilare con: g++ -std=c++11 esercizio20.cpp
*/

#include <iostream>
#include <functional>

using std::cout;
using std::endl;
using std::function;

int main() {
	// funzione "successore"

	function<int(int)> s = [] (const int n) { return n + 1; };

	// restituisce il successore del successore di un numero

	function<int(function<int(int)>, int)> ff = [] (function<int(int)> f, const int n) { return f(f(n)); };

	// stampa "2"

	cout << "Successore del successore di zero: " << ff(s, 0) << endl;
}

